sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala

```diff
@@ -3286,6 +3265,25 @@ class Analyzer(override val catalogManager: CatalogManager) extends RuleExecutor
     }
   }
 
+  /**
+   * Resolves expressions against their scope of attributes.
+   */
+  object ResolveScopedExpression extends Rule[LogicalPlan] {
+    override def apply(plan: LogicalPlan): LogicalPlan = plan.resolveExpressions {
+      case se: ScopedExpression if se.resolved => se.child
+      case se @ ScopedExpression(expr, attributes) =>
```

Comment:
one thing is we need to make sure `attributes` is a subset of the children output.

Reply:
> we need to make sure `attributes` is a subset of the children output.

I think this only holds for `AttributeReference`, because `UnresolvedAttribute` that cannot be resolved against children output will cause an Analyzer error.

Reply:
a lambda can match the first child of `CoGroup` and apply special logic if it's `AppendColumns`, another lambda can match the second child of `CoGroup`.

Reply:
> apply special logic if it's AppendColumns,

Sounds like you still want to bypass the children if it is a `AppendColumns`, as we used to do in (and few other places):
```
object ResolveReferencesInAggregate extends SQLConfHelper
  with ColumnResolutionHelper with AliasHelper {
  def apply(a: Aggregate): Aggregate = {
    val planForResolve = a.child match {
      // SPARK-25942: Resolves aggregate expressions with `AppendColumns`'s children, instead of
      // `AppendColumns`, because `AppendColumns`'s serializer might produce conflict attribute
      // names leading to ambiguous references exception.
      case appendColumns: AppendColumns => appendColumns
      case _ => a
    }
...
```

We cannot know whether the child of a `MapGroups` / `CoGroup` is an `AppendColumns` that adds the key column (which we want to bypass or ignore), or an `AppendColumns` that comes from `mapValues` (which we do not want to ignore). See the _"However, mapValues introduces another 'value' column, ..."_ section in the PR description above.